### PR TITLE
UCP/PROTO/CORE: Minor improvements for Zcopy code

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

1. `ucp_am_zcopy_complete_last_stage` -> `ucp_request_zcopy_complete_last_stage` and accepts proto ID and status parameters
2. don't check for error status in `ucp_request_send_state_advance`, use `assert()` instead, it accepts only `UCS_INPROGRESS` and `UCS_OK`
3. reorganize code in `ucp_do_am_zcopy_single` to remove extra branches and use common `ucp_request_zcopy_complete_last_stage` function to handle `UCS_INPROGRESS` and `UCS_OK` return values
4. improvements in `ucp_do_am_zcopy_multi` function:
a) added missing braces
b) reorganize code for the last stage handling - just call `ucp_request_zcopy_complete_last_stage` if 
`UCS_INPROGRESS` or `UCS_OK` and no need for getting the next lane
c) simplify code for the handling `N-1`th segments (where N -the number of segments) - if the error was returned, handle if there are no resource and return from the function
5. update `ucp_rma_basic_progress_put` function to advance request state only in case of the error
6. reorganize code in `ucp_rndv_progress_rma_get_zcopy` and `ucp_rndv_progress_rma_put_zcopy` functions to advance only in case of `UCS_OK` or `UCS_INPROGRESS`

## Why ?

1. to use this common code in more places
2. to save this branch in some places in `ucp_do_am_zcopy_multi` functions
3. to use common code and reduce the number of branches
4. to simplify code
5. to be able use `ucp_request_send_state_advance` that don't accept error status anymore
6. to use `ucp_request_zcopy_complete_last_stage` instead of duplicated code

## How ?

1. just renaming and adding new parameters
2. replace `if` statement and `return` inside the block by `assert()`
3. use `ucp_request_zcopy_complete_last_stage` if UCT AM copy reutnrs either `UCS_OK` or `UCS_INPROGRESS`
4. remove extra `if`
5. add additional `if` to check for the status
6. check for the status from the UCT PUT/GET Zcopy and if is `UCS_OK`/`UCS_INPROGRESS`, handle the processed segment

### CODE REVIEW

**Comment by brminich** [comment]
can you plz describe the fixed use-case?

**Comment by dmitrygx** [comment]
> can you plz describe the fixed use-case?

sure, update `WHAT` section in the description to make it clear

**Comment by brminich** [comment]
In this case you will miss `advance` for `NO_RESOURCE` case (which is quite likely to occur), right?

**Comment by brminich** [comment]
btw `ucp_request_send_state_advance` does nothing in case of error, so this is not a real problem

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@brminich do you think this PR makes sense?
for me it makes Zcopy handling clear and fixes some extra actions (removing switching to the next lane when this is not needed, calling advancing state procedure twice in case of the error - it does nothing, but looks unclear)

File: src/ucp/core/ucp_request.inl [hunk 1]
```diff
@@ -384,6 +382,24 @@ ucp_request_send_state_advance(ucp_request_t *req,
...
 
+static UCS_F_ALWAYS_INLINE void
+ucp_request_zcopy_complete_last_stage(ucp_request_t *req, ucp_dt_state_t *state,
+                                      unsigned proto, ucs_status_t status,
+                                      ucp_req_complete_func_t complete)
+{
+    ucs_assert(!UCS_STATUS_IS_ERR(status));
+    ucp_request_send_state_advance(req, state, proto, status);
+
+    /* Complete a request on a last stage if all previous AM
```
Comment (brminich) [inline src/ucp/core/ucp_request.inl:393]: it is called from `put` and `get` as well

File: src/ucp/proto/proto_am.inl [hunk 2]
```diff
@@ -346,9 +327,11 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
                 status = uct_ep_am_zcopy(uct_ep, am_id_middle, (void*)hdr_middle,
                                          hdr_size_middle, iov, iovcnt, 0,
                                          &req->send.state.uct_comp);
-            } else if (state.offset == req->send.length) {
+            } else if (req->send.state.dt.offset == req->send.length) {
```
Comment (brminich) [inline src/ucp/proto/proto_am.inl:330]: why change to `req->send.state`?

Comment (dmitrygx) [reply src/ucp/proto/proto_am.inl:330]: returned back
doesn't matter which value is used here

**Comment by dmitrygx** [comment]
@brminich @yosefe should I close this PR? or maybe you will evaluate it?

**Comment by brminich** [comment]
please, close it

**Pull request closed without merging**